caffe2/onnx/onnx_exporter.cc

```diff
@@ -0,0 +1,585 @@
+  TensorProto tensor;
+  tensor.set_name(DummyName::NewDummyName());
+  tensor.set_data_type(TensorProto::INT64);
+  tensor.add_dims(shape.size());
+  tensor.mutable_raw_data()->assign(
+      reinterpret_cast<const char*>(shape.data()), sizeof(int64_t) * shape.size());
+  return tensor;
+}
+} // namespace
+
+const std::unordered_map<std::string, std::string>&
+OnnxExporter::get_renamed_operators() const {
+  const static std::unordered_map<std::string, std::string> kRenamedOperators{
+      {"SpatialBN", "BatchNormalization"},
+      {"Conv1D", "Conv"},
+      {"Conv2D", "Conv"},
+      {"Conv3D", "Conv"},
+      {"ConvTranspose1D", "ConvTranspose"},
+      {"ConvTranspose2D", "ConvTranspose"},
+      {"ConvTranspose3D", "ConvTranspose"},
+      {"MaxPool1D", "MaxPool"},
+      {"MaxPool2D", "MaxPool"},
+      {"MaxPool3D", "MaxPool"},
+      {"AveragePool1D", "AveragePool"},
+      {"AveragePool2D", "AveragePool"},
+      {"AveragePool3D", "AveragePool"}};
+  return kRenamedOperators;
+}
+
+const std::unordered_map<std::string, std::string>&
+OnnxExporter::get_renamed_attrs() const {
+  const static std::unordered_map<std::string, std::string> kRenamedAttrs{
+      {"kernels", "kernel_shape"}};
+  return kRenamedAttrs;
+}
+
+const std::
+    unordered_map<std::string, std::unordered_map<std::string, std::string>>&
+    OnnxExporter::get_per_op_renamed_attrs() const {
+  const static std::
+      unordered_map<std::string, std::unordered_map<std::string, std::string>>
+          kPerOpRenamedAttrs = {{"Squeeze", {{"dims", "axes"}}},
+                                {"Unsqueeze", {{"dims", "axes"}}},
+                                {"Transpose", {{"axes", "perm"}}},
+                                {"ConvTranspose", {{"adjs", "output_padding"}}},
+                                {"Selu", {{"scale", "gamma"}}}};
+
+  return kPerOpRenamedAttrs;
+}
+
+const std::unordered_map<std::string, OnnxExporter::SpecialOpConverter>&
+OnnxExporter::get_special_operators() const {
+  const static std::unordered_map<std::string, OnnxExporter::SpecialOpConverter>
+      kSpecialOperators = {
+        {"Conv", &OnnxExporter::CreateConvPoolNodes},
+        {"ConvTranspose", &OnnxExporter::CreateConvPoolNodes},
+        {"MaxPool", &OnnxExporter::CreateConvPoolNodes},
+        {"AveragePool", &OnnxExporter::CreateConvPoolNodes},
+        {"FC", &OnnxExporter::CreateGemmNodes},
+        {"Concat", &OnnxExporter::CreateConcatNodes},
+        {"LRN", &OnnxExporter::CreateLrnNodes},
+        {"Reshape", &OnnxExporter::CreateReshapeNodes},
+        {"Slice", &OnnxExporter::CreateSliceNodes},
+        {"ChannelShuffle",  &OnnxExporter::CreateChannelShuffleNodes}
+      };
+  return kSpecialOperators;
+}
+
+void OnnxExporter::CopyCaffe2ArgToOnnxAttr(
+    AttributeProto* attr,
+    const std::string& op_type,
+    const caffe2::Argument& arg) {
+  std::string name;
+  if (get_per_op_renamed_attrs().count(op_type)) {
```

Comment:
If you use the value when its present, it is better to do `std::map<T>.find()` and compare the result to `std::map<T>.end()` to find if the key is present in the map. If it is, you can rename the iterator returned by `std::map<T>.find()`. Thus you will search the map only once, not twice.

Reply:
Got it. I noticed this but since the map is very small I didn't bother. I can change it back.